courses/advanced-ada/chapters/generics.rst

```diff
@@ -1084,19 +1086,335 @@ In this example, we declare the array :ada:`A` with elements of both
+to deduce that the intention of declaring :ada:`T` and :ada:`Hash` is to
+define an interface. However, this approach has the advantage of not
+requiring the use of tagged types in the package instantiation.
+
+This is an example of a package instantiating the generic hash table:
+
+.. code:: ada
+
+    with Ada.Containers; use Ada.Containers;
+    with Ada.Strings.Hash;
+
+    with Interface_Using_Formal_Function;
+    use  Interface_Using_Formal_Function;
+
+    package Instantiation_Using_Formal_Function is
+
+       type My_Type is record
+          Key   : String (1 .. 100);
+          Key_2 : String (1 .. 100);
+       end record;
+
+       function Hash (Self : My_Type) return Hash_Type is
+         (Ada.Strings.Hash (Self.Key));
+
+       function Alt_Hash (Self : My_Type) return Hash_Type is
+         (Ada.Strings.Hash (Self.Key_2));
+
+       package My_Type_Hash_Tables is new Hash_Tables (My_Type);
+
+       package My_Type_Alt_Hash_Tables is new Hash_Tables
+         (T    => My_Type,
+          Hash => Alt_Hash);
+
+    end Instantiation_Using_Formal_Function;
+
+Note that, in the declaration of the :ada:`My_Type_Hash_Tables`, we're
+not specifying the :ada:`Hash` function for the instantiation of the
+generic :ada:`Hash_Tables` package. This is possible for two reasons:
+
+- In the declaration of the formal function parameter, we're using
+  :ada:`is <>`, which automatically selects a function with the same name
+  in the package instantiation if available.
+
+- For :ada:`My_Type`, we've declared a function that has the same name as
+  the formal function.
+
+If the above-mentioned conditions are not met, we have to provide an
+argument for the formal function parameter in the package instantiation.
+
+We may also instantiate the formal package using alternative versions of
+the function associated with the formal package. This is what we're doing
+in the declaration of the :ada:`My_Type_Alt_Hash_Tables` package. In this
+case, we're using :ada:`Alt_Hash` instead of :ada:`Hash` for the formal
+function parameter. Note that, because the name of the actual function
+doesn't match the name of the formal function, we need to indicate it
+explicitly.
+
+Interfaces using signature packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Signature packages are used to group a set of types and subprograms that
+serve as a formal package parameter in another generic package. The basic
+form is similar to the approach we've just seen using formal subprograms:
+a signature package defines an interface using a formal type and formal
+subprograms.
+
+Signature packages make it more explicit that the types and subprograms
+defined in the package represent an interface. This is an advantage over
+the approach using formal subprograms directly. However, using signature
+package isn't as explicit as using the :ada:`interface` keyword.
+
+In general, signature packages aren't used in isolation, but in
```

Comment:
another strong hint is that in general signature packages don't define anything themselves

Reply:
Yep, related to my comment